File: MergeSort.cpp
// Merge takes time O(n)
// Whole algorithm, worse case O(log(n))
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

void Merge(vector<int>& vec, int p, int q, int r) {
  int len_1 = q - p + 1;
  int len_2 = r - q;
  vector<int> lhs(len_1 + 1, 0);
  vector<int> rhs(len_2 + 1, 0);
  for(int i = 0; i < len_1; ++i) {
    lhs[i] = vec[i + p];
  }
  for(int j = 0; j < len_2; ++j) {
    rhs[j] = vec[j + q + 1];
  }

  lhs[len_1] = INT_MAX;
  rhs[len_2] = INT_MAX;

  int i = 0;
  int j = 0;

  for(int k = p; k <= r; ++k) {
    if (lhs[i] < rhs[j]) {
      vec[k] = lhs[i];
      ++i;
    }
    else {
      vec[k] = rhs[j];
      ++j;
    }
  }
}

void MergeSort(vector<int>& vec, int p, int r) {
  // It's very important to check if p < r!!!
  if (p < r) {
    int q = (p + r) / 2;
    // 0 to q; q+1 to r
    MergeSort(vec, p, q);
    MergeSort(vec, q + 1, r);
    Merge(vec, p, q, r);
  }
}

void Print(vector<int>& vec) {
  for(auto v : vec) {
    cout << v << " ";
  }
  cout << endl;
}

int main() {
  vector<int> vec{2037, -100, 0, 999, 12, 66, 5, 999};
  cout << "Pre-MergeSort: " << endl;
  Print(vec);
  MergeSort(vec, 0, vec.size() - 1);
  cout << "Post-MergeSort: " << endl;
  Print(vec);
  return 0;
}

File: MergeSort.py
def Merge(arr, p, q, r):
    l_size = q - p + 1
    r_size = r - q

    arr_left = arr[p:q+1]
    arr_right = arr[q+1:r+1]

    i = 0
    j = 0
    k = p

    while (i < l_size) and (j < r_size):
        if (arr_left[i] > arr_right[j]):
            arr[k] = arr_right[j]
            j += 1
        else:
            arr[k] = arr_left[i]
            i += 1
        k += 1

    while(i < l_size):
        arr[k] = arr_left[i]
        i += 1
        k += 1

    while(j < r_size):
        arr[k] = arr_right[j]
        j += 1
        k += 1

def MergeSort(arr, p, r):
    if (p < r):
        q = (int)((p + r) / 2)
        MergeSort(arr, p, q)
        MergeSort(arr, q + 1, r)
        Merge(arr, p, q, r)

arr = [2037, -100, -66, 999, 0, 15, 22, 1024, -66, 10101010, 234, -234]
size = len(arr)
print("Pre-MergeSort: \n", arr)
MergeSort(arr, 0, size - 1)
print("Post-MergeSort: \n", arr)
